### ISSUE TITLE
builtin pollutes global namespace

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#4733]
It seems that the `builtin` part of Minetest pretty boldly pollutes the global namespace of Lua (`_G`) with lots of variables. I have figured this out by using the mod `qa_block` (https://forum.minetest.net/viewtopic.php?f=11&t=15759&p=238864) and some research.

Anyway, by using this mod, I have determined that the following variables are in global namespace after loading Minetest even with just the most basic set of mods (I tested it with an empty subgame and only 2 mods: `origin` and `qa_blocks`):

``` 
   cleanup_path
   gcinfo
   on_placenode
   hack_nodes
   file_exists
   nodeupdate
   check_attached_node
   drop_attached_node
   newproxy
   get_last_folder
   spawn_falling_node
   on_dignode
   basic_dump
   nodeupdate_single
   INIT
   DIR_DELIM
   PLATFORM
```

I am pretty sure they come from `builtin`, and not some other mod or Lua itself. For some variables, I could directly verify its existance in `builtin`, but I have not checked every single variable (reason: Laziness :D).
Long story short, `builtin` is in clear violation of the Lua style guide, which says “Avoid globals like the plague.” ;-)
The problem is, the variables above leak into the namespace which is also used by ordinary Minetest mods. It can probably lead to trouble if, by chance, a modder uses any of the names above (and thus overwrites it).

Most variables above are functions. The all-caps variables are strings. Please verify for yourselves if these are actually from `builtin`, and not actually part of the Lua API.

I think the only solution to this is by putting all these variables into a “namespace” table or something like this. I don't see any single reason why so many variables should be in `_G`.

**Comment by sofar** [comment]
Many of these are plainly absent from any API documentation and are only used locally. We should immediately make them local as soon as we can.

Exceptions I can see for now:
- INIT - used by the menu mod system
- nodeupdate (deprecated, but used) and nodeupdate_single
- DIR_DELIM

those need API documentation IMO.

**Comment by Wuzzy2** [comment]
If you are going to make `nodeupdate` and `nodeupdate_single` an officially documented part of the API, they should definitely go into the `minetest` table, of course.

I agree that `DIR_DELIM` is useful outside of builtin. I don't really understand the other variables.

**Comment by est31** [comment]
For `nodeupdate` there is issue #3639 . I  agree that if its part of the official API, it should be put into `minetest`.

**Comment by paramat** [comment]
> I agree that if its part of the official API, it should be put into minetest.

I'm planning to do this, i assume it's as simple as adding 'core.' before the function?
I'll change the docs too.

**Comment by Wuzzy2** [comment]
5.1.1 update: Most unwanted globals are gone. Remaining are these functions:

* `cleanup_path`
* `file_exists`
* `get_last_folder`
* `basic_dump`

And `INIT` and `DIR_DELIM` need documentation. I understand the meaning of `DIR_DELIM` but what does `INIT` do and is it of any use for modders?

**Comment by Wuzzy2** [comment]
Hmm, I'm thinking about `INIT` and `DIR_DELIM` again …
Do modders really need to know about `INIT` and `DIR_DELIM` anyway?

I never used `DIR_DELIM` when I load a file via `dofile` and happily just hardcode normal slashes inside, yet nobody ever complained my mods break on Windows …
Apparently the normal slash just works everywhere.

And `INIT` seems to be only relevant for builtin.

**Comment by rubenwardy** [comment]
Dir delim should be deprecated. Mods should use /, with the API translating to \ where needed - Lua already does this in its standard library

I believe that init is used to say whether the lua environment is server, client, or mainmenu

**Comment by Wuzzy2** [comment]
Note that `DIR_DELIM` and `INIT` were never part of the official API, so any changes are safe.

It seems those two are defined in C++, not in Lua builtin code. Maybe the C++ code neds to be adapted to only make `DIR_DELIM` and `INIT` visible to builtin, but not for mods.

I don't know how it really works. Does builtin share the namespace with mods?

**Comment by sfan5** [comment]
> Note that DIR_DELIM and INIT were never part of the official API, so any changes are safe.

"Safe" is relative. If you remove something that 50% of mods use, then we have a problem even if the thing was undocumented.

> Does builtin share the namespace with mods?

Yes.

**Comment by Wuzzy2** [comment]
#9451 was merged.
Now the only things in the list that remain are `INIT` and `DIR_DELIM`.

Question: Should we document those or just leave them be?

If it's the latter, then this issue can be closed.

**Comment by appgurueu** [comment]
I'm against `local`izing useful functions. I would rather recommend adding those helpful functions to the API & documenting them.

**Comment by Desour** [comment]
- `INIT` can be used to use one lua file in different places. (Eg. a server mod and a csm both dofile a common base.)
- `DIR_DELIM` is documented on the dev wiki (by me, sorry): https://dev.minetest.net/DIR_DELIM
- Functions like `file_exists` aren't helpful, mods shouldn't play around with files but use `minetest.get_mod_storage()` instead.

**Comment by rubenwardy** [comment]
Using / is a lot better than DIR_DELIM. The lua standard library automatically converts / to \ on Windows, and so should Minetest (it probably mostly does this already). The benefit of Lua is that you don't need to care what the platform is or the bad decisions Microsoft made - you are isolated from it

**Comment by Wuzzy2** [comment]
Given that both `DIR_DELIM` and `INIT` are not meant for mod use, I consider this issue done. The global namespace pollusion is gone.

Closing.

---

Off topic:

> The Dev wiki doesn't count as official documentation, and so we don't have to maintain compatibility with undocumented functions that are documented on the wiki

Yes, it's true, but for unsuspecting people the wiki VERY much looks like official documentation. It really shouldn't be this way. I have already sprinkled warning signs here and there, but that's not good enough. The problem is that there is a huge number of pages that just copied the function documentation from years ago.
The real solution would be to mass-delete all >200 pages in the “Methods” category, and most other pages that were just ripped from `lua_api.txt`. But this is easier said than done in MediaWiki.
Maybe someone with admin powers can do it, but the admin seems to be never available. :(
